src/humanloop/eval_utils/run.py

```diff
@@ -0,0 +1,751 @@
+                # TODO: support the case where `file` logs generated on Humanloop but Evaluator logs generated locally
+                if function_ is None:
+                    raise ValueError(
+                        "Local Evaluators are only supported when generating Logs locally using your "
+                        f"{type_}'s `callable`. Please provide a `callable` for your file in order "
+                        "to run Evaluators locally."
+                    )
+                local_evaluators.append(evaluator)
+                spec = ExternalEvaluator(
+                    arguments_type=evaluator["args_type"],
+                    return_type=evaluator["return_type"],
+                    attributes={"code": inspect.getsource(eval_function)},
+                    evaluator_type="external",
+                )
+                client.evaluators.upsert(
+                    id=evaluator.get("id"),
+                    path=evaluator.get("path"),
+                    spec=spec,
+                )
+    function_ = typing.cast(Callable, function_)
+
+    # Validate upfront that the local Evaluators and Dataset fit
+    requires_target = False
+    for local_evaluator in local_evaluators:
+        if local_evaluator["args_type"] == "target_required":
+            requires_target = True
+            break
+    if requires_target:
+        missing_target = 0
+        for datapoint in hl_dataset.datapoints:  # type: ignore
+            if not datapoint.target:
+                missing_target += 1
+        if missing_target > 0:
+            raise ValueError(
+                f"{missing_target} Datapoints have no target. A target "
+                f"is required for the Evaluator: {local_evaluator['path']}"
+            )
+
+    # Get or create the Evaluation based on the name
+    evaluation = None
+    try:
+        evaluation = client.evaluations.create(
+            name=name,
+            evaluators=[{"path": e["path"]} for e in evaluators],  # type: ignore
+            file={"id": hl_file.id},
+        )
+    except ApiError as error_:
+        # If the name exists, go and get it # TODO: Update API GET to allow querying by name and file.
+        if error_.status_code == 409:
+            evals = client.evaluations.list(file_id=hl_file.id, size=50)
+            for page in evals.iter_pages():
+                evaluation = next((e for e in page.items if e.name == name), None)  # type: ignore
+        else:
+            raise error_
+        if not evaluation:
+            raise ValueError(f"Evaluation with name {name} not found.")
+
+    # Create a new Run
+    run: EvaluationRunResponse = client.evaluations.create_run(
+        id=evaluation.id,
+        dataset={"version_id": hl_dataset.version_id},
+        orchestrated=False,
+    )
+    # Every Run will generate a new batch of Logs
+    run_id = run.id
+
+    _PROGRESS_BAR = _SimpleProgressBar(len(hl_dataset.datapoints))  # type: ignore
+
+    if is_decorated:
```

Comment:
what difference is there in the logic between decorated and non-decorated functions? (hard to parse looking at this code; would appreciate a comment here describing this)

my guess is that in the non-decorated case we call log_func?

shouldn't we call log_func in both cases and perform a check as to whether we have already logged to this run and datapoint pair for this file?
(thought we discussed this in office)

Reply:
Agree - changed

Reply:
Code seems the same here, is the change somewhere else? Noting that I made some previous comments around removing `is_decorated` as no longer needed because I had read this was changed in passing.

Reply:
you are commenting on an outdated version of the file